src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/ConstantExpectedAnalyzer.cs

```diff
@@ -0,0 +1,566 @@
+
+        protected void OnParameterWithConstantExpectedAttribute(IParameterSymbol parameter, Action<Diagnostic> reportAction)
+        {
+            if (!ValidateConstantExpectedParameter(parameter, out ImmutableArray<Diagnostic> diagnostics))
+            {
+                foreach (var diagnostic in diagnostics)
+                {
+                    reportAction(diagnostic);
+                }
+            }
+        }
+
+        private static bool TryCreateConstantExpectedParameter(IParameterSymbol parameterSymbol, [NotNullWhen(true)] out ConstantExpectedParameter? parameter)
+        {
+            if (!TryGetConstantExpectedAttributeData(parameterSymbol, out var attributeData))
+            {
+                parameter = null;
+                return false;
+            }
+
+            switch (parameterSymbol.Type.SpecialType)
+            {
+                case SpecialType.System_Char:
+                    return UnmanagedHelper<char>.TryCreate(parameterSymbol, attributeData, char.MinValue, char.MaxValue, out parameter);
+                case SpecialType.System_Byte:
+                    return UnmanagedHelper<ulong>.TryCreate(parameterSymbol, attributeData, byte.MinValue, byte.MaxValue, out parameter);
+                case SpecialType.System_UInt16:
+                    return UnmanagedHelper<ulong>.TryCreate(parameterSymbol, attributeData, ushort.MinValue, ushort.MaxValue, out parameter);
+                case SpecialType.System_UInt32:
+                    return UnmanagedHelper<ulong>.TryCreate(parameterSymbol, attributeData, uint.MinValue, uint.MaxValue, out parameter);
+                case SpecialType.System_UInt64:
+                    return UnmanagedHelper<ulong>.TryCreate(parameterSymbol, attributeData, ulong.MinValue, ulong.MaxValue, out parameter);
+                case SpecialType.System_UIntPtr:
+                    return UnmanagedHelper<ulong>.TryCreate(parameterSymbol, attributeData, uint.MinValue, uint.MaxValue, out parameter);
+                case SpecialType.System_SByte:
+                    return UnmanagedHelper<long>.TryCreate(parameterSymbol, attributeData, sbyte.MinValue, sbyte.MaxValue, out parameter);
+                case SpecialType.System_Int16:
+                    return UnmanagedHelper<long>.TryCreate(parameterSymbol, attributeData, short.MinValue, short.MaxValue, out parameter);
+                case SpecialType.System_Int32:
+                    return UnmanagedHelper<long>.TryCreate(parameterSymbol, attributeData, int.MinValue, int.MaxValue, out parameter);
+                case SpecialType.System_Int64:
+                    return UnmanagedHelper<long>.TryCreate(parameterSymbol, attributeData, long.MinValue, long.MaxValue, out parameter);
+                case SpecialType.System_IntPtr:
+                    return UnmanagedHelper<long>.TryCreate(parameterSymbol, attributeData, int.MinValue, int.MaxValue, out parameter);
```

Comment:
@tannergooding In your opinion, is it a good practice to check for int.MinValue and int.MaxValue, since IntPtr has a platform-dependent size?

Reply:
See https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nint-nuint#constants

Reply:
Right. But these constants only affect 32-bit systems. On 64-bit systems they would be long.MaxValue and ulong.MaxValue respectively.

The question is whether we want to make the maximums platform-dependent or go platform-independent to 32-bit.

Reply:
But what you can set as a constant is fixed.
See https://github.com/dotnet/docs/issues/24287